daemon/containerd/image_delete.go

```diff
@@ -78,9 +140,179 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
 		logrus.WithError(err).Warn("failed to unlease snapshots")
 	}
 
-	imgID := string(img.Target.Digest)
-	i.LogImageEvent(imgID, imgID, "untag")
-	i.LogImageEvent(imgID, imgID, "delete")
+	imgIDStr := img.Target.Digest.String()
+	i.LogImageEvent(imgIDStr, imgIDStr, "untag")
+	i.LogImageEvent(imgIDStr, imgIDStr, "delete")
 
-	return []types.ImageDeleteResponseItem{{Untagged: reference.FamiliarString(parsedRef)}}, nil
+	return records, nil
+}
+
+// isImageIDPrefix returns whether the given possiblePrefix is a prefix of the
+// given imageID.
+func isImageIDPrefix(imageID, possiblePrefix string) bool {
+	if strings.HasPrefix(imageID, possiblePrefix) {
+		return true
+	}
+
+	if i := strings.IndexRune(imageID, ':'); i >= 0 {
+		return strings.HasPrefix(imageID[i+1:], possiblePrefix)
+	}
+
+	return false
+}
+
+// isSingleReference returns true if there is no other image
+// with the same target, not including dangling images
+func (i *ImageService) isSingleReference(ctx context.Context, imgID image.ID) bool {
+	imgs, err := i.client.ImageService().List(ctx, "target.digest=="+imgID.String())
+	if err != nil {
+		return true
+	}
+	var notDangling int
+	for _, img := range imgs {
+		if !isDanglingImage(img) {
+			notDangling++
+		}
+	}
+	return notDangling == 1
+}
+
+// ImageDeleteConflict holds a soft or hard conflict and an associated error.
+// Implements the error interface.
+type imageDeleteConflict struct {
+	hard    bool
+	used    bool
+	imgID   image.ID
+	message string
+}
+
+func (idc *imageDeleteConflict) Error() string {
+	var forceMsg string
+	if idc.hard {
+		forceMsg = "cannot be forced"
+	} else {
+		forceMsg = "must be forced"
+	}
+
+	return fmt.Sprintf("conflict: unable to delete %s (%s) - %s", stringid.TruncateID(idc.imgID.String()), forceMsg, idc.message)
+}
+
+func (idc *imageDeleteConflict) Conflict() {}
+
+// imageDeleteHelper attempts to delete the given image from this daemon. If
+// the image has any hard delete conflicts (child images or running containers
+// using the image) then it cannot be deleted. If the image has any soft delete
+// conflicts (any tags/digests referencing the image or any stopped container
+// using the image) then it can only be deleted if force is true. If the delete
+// succeeds and prune is true, the parent images are also deleted if they do
+// not have any soft or hard delete conflicts themselves. Any deleted images
+// and untagged references are appended to the given records. If any error or
+// conflict is encountered, it will be returned immediately without deleting
+// the image. If quiet is true, any encountered conflicts will be ignored and
+// the function will return nil immediately without deleting the image.
+func (i *ImageService) imageDeleteHelper(ctx context.Context, imgID image.ID, records *[]types.ImageDeleteResponseItem, force, prune, quiet bool) error {
```

Comment:
I'll look into it! I'm not sure if a bitmask will help readability (esp. since conflict types are already implemented as a bitmask, and force is interacting with that, but maybe there's a better way of modeling that together – or just a config struct with clearly named parameters)

Reply:
I cleaned things up a bit and we don't have multiple bool flag arguments getting passed around anymore 🥳